crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1166,18 +1179,138 @@ impl<'db> TypeInferenceBuilder<'db> {
         let result = infer_expression_types(self.db, expression);
         self.extend(result);
-        let value_ty = self.expression_ty(&assignment.value);
-        self.add_binding(assignment.into(), definition, value_ty);
+
+        let value_ty = self.expression_ty(value);
+
+        let target_ty = match (value_ty, kind) {
+            (
+                Type::Tuple(_) | Type::StringLiteral(_) | Type::LiteralString,
+                AssignmentKind::Sequence,
+            ) => self.infer_sequence_unpacking(target, value_ty, variable),
+            (_, AssignmentKind::Sequence) => value_ty
+                .iterate(self.db)
+                .unwrap_with_diagnostic(value.into(), self),
+            _ => value_ty,
+        };
+
+        self.add_binding(variable.into(), definition, target_ty);
         self.types
             .expressions
-            .insert(target.scoped_ast_id(self.db, self.scope), value_ty);
+            .insert(variable.scoped_ast_id(self.db, self.scope), target_ty);
+    }
+
+    fn infer_sequence_unpacking(
+        &mut self,
+        target: &ast::Expr,
+        value_ty: Type<'db>,
+        variable: &ast::ExprName,
+    ) -> Type<'db> {
+        // The inner function is recursive and only differs in the return type which is an `Option`
+        // where if the variable is found, the corresponding type is returned otherwise `None`.
+        fn inner<'db>(
+            builder: &mut TypeInferenceBuilder<'db>,
+            target: &ast::Expr,
+            value_ty: Type<'db>,
+            variable: &ast::ExprName,
+        ) -> Option<Type<'db>> {
+            match target {
+                // TODO: Handle other possible target expression like attribute, subscript,
+                // and starred.
+                ast::Expr::Name(name) if name == variable => {
+                    return Some(value_ty);
+                }
+                ast::Expr::List(ast::ExprList { elts, .. })
+                | ast::Expr::Tuple(ast::ExprTuple { elts, .. }) => match value_ty {
+                    Type::Tuple(tuple_ty) => {
+                        let starred_index = elts.iter().position(ast::Expr::is_starred_expr);
+
+                        let element_types = if let Some(starred_index) = starred_index {
+                            if tuple_ty.len(builder.db) >= elts.len() {
+                                let mut element_types = Vec::with_capacity(elts.len());
+                                element_types.extend_from_slice(
+                                    // SAFETY: Safe because of the length check above.
+                                    tuple_ty.elements(builder.db).get(..starred_index).unwrap(),
+                                );
+
+                                // E.g., in `(a, *b, c, d) = ...`, the index of starred element `b`
+                                // is 1 and the remaining elements after that are 2.
+                                let remaining = elts.len() - 2;
```

Comment:
This doesn't look right to me; the `- 2` seems hardcoded to the case where `starred_index == 1`, and all the starred unpacking tests also have `starred_index == 1`. If I change one of the tests to have the starred element in a different position, inference of the element types from the end doesn't seem to work correctly.

If this variable is supposed to be the number of remaining elements in the unpacking target, it seems like that should be something like `elts.len() - (starred_index + 1)` instead?

And let's make sure we add a test for starred unpack where the starred element is not in second position.

Reply:
Yes, thanks for catching that, I totally did not intend that to happen.